File: android/src/main/java/com/omnix/voice/OmnixVoice.kt
package com.omnix.voice

import android.content.Context
import android.util.Log
import com.omnix.voice.internal.OmnixEventDispatcher
import com.omnix.voice.internal.OmnixNative
import com.omnix.voice.internal.OmnixNativeCallback
import com.omnix.voice.internal.OmnixPermissions
import java.util.concurrent.ConcurrentHashMap

/**
 * Public Omnix Voice singleton (Issue #1 §20A.1 / SDK-032).
 *
 * No `external` / JNI types. Baresip/re types must never appear here.
 * Password is passed to native once and is never persisted by this SDK.
 * [initialize] requires [android.Manifest.permission.RECORD_AUDIO] (SDK-033).
 */
object OmnixVoice {
    private const val TAG = "OmnixVoice"
    private const val NATIVE_OK = 0

    private val lock = Any()
    private val dispatcher = OmnixEventDispatcher()
    private val calls = ConcurrentHashMap<String, OmnixCallInfo>()

    @Volatile
    private var appContext: Context? = null

    @Volatile
    private var initialized: Boolean = false

    @Volatile
    private var audioRoute: OmnixAudioRoute = OmnixAudioRoute.UNKNOWN

    @Volatile
    var registrationState: OmnixRegistrationState =
        OmnixRegistrationState.UNINITIALIZED
        private set

    private val nativeCallback =
        object : OmnixNativeCallback {
            override fun onRegistrationStateChanged(
                state: Int,
                sipCode: Int,
                reason: String?,
            ) {
                val mapped = OmnixRegistrationState.fromNative(state)
                registrationState = mapped
                val codeOrNull = if (sipCode == 0) null else sipCode
                dispatcher.dispatch {
                    it.onRegistrationStateChanged(mapped, codeOrNull, reason)
                }
            }

            override fun onCallEvent(
                state: Int,
                callId: String?,
                peerUri: String?,
                peerDisplayName: String?,
                isOutgoing: Boolean,
                isMuted: Boolean,
                isOnHold: Boolean,
            ) {
                val id = callId ?: return
                val mapped = OmnixCallState.fromNative(state)
                val info =
                    OmnixCallInfo(
                        callId = id,
                        peerUri = peerUri.orEmpty(),
                        peerDisplayName = peerDisplayName.orEmpty(),
                        state = mapped,
                        isOutgoing = isOutgoing,
                        isMuted = isMuted,
                        isOnHold = isOnHold,
                    )
                calls[id] = info
                if (mapped == OmnixCallState.INCOMING) {
                    dispatcher.dispatch { it.onIncomingCall(info) }
                }
                dispatcher.dispatch { it.onCallStateChanged(info) }
                if (mapped == OmnixCallState.ENDED || mapped == OmnixCallState.FAILED) {
                    // Keep last snapshot for getCallInfo briefly; remove idle noise later.
                }
            }

            override fun onError(code: Int, detail: String?) {
                val mapped = OmnixErrorCode.fromNative(code)
                // Never log detail if it could contain secrets; facade detail is opaque.
                Log.w(TAG, "native error code=$mapped")
                dispatcher.dispatch { it.onError(mapped, detail, null) }
            }
        }

    /**
     * Initialize the SDK. Stores [Context.getApplicationContext] only.
     * Does not write [OmnixVoiceConfig.sipPassword] to disk.
     *
     * Requires [android.Manifest.permission.RECORD_AUDIO] already granted.
     * If missing, fires [OmnixVoiceListener.onError] with
     * [OmnixErrorCode.PERMISSION_DENIED] and throws [OmnixVoiceException].
     */
    @JvmStatic
    fun initialize(context: Context, config: OmnixVoiceConfig) {
        synchronized(lock) {
            if (initialized) {
                throw OmnixVoiceException(
                    OmnixErrorCode.INVALID_CALL_STATE,
                    "already initialized",
                )
            }
            if (config.sipServer.isBlank() || config.sipUser.isBlank()) {
                throw OmnixVoiceException(
                    OmnixErrorCode.INVALID_CONFIGURATION,
                    "sipServer and sipUser are required",
                )
            }

            // SDK-033: check only — host must request the runtime permission.
            if (!OmnixPermissions.isRecordAudioGranted(context)) {
                val detail = OmnixPermissions.RECORD_AUDIO_DETAIL
                dispatcher.dispatch {
                    it.onError(OmnixErrorCode.PERMISSION_DENIED, detail, null)
                }
                OmnixPermissions.requireRecordAudioGranted(false)
            }

            appContext = context.applicationContext
            OmnixNative.nativeSetCallback(nativeCallback)

            val codecsCsv =
                config.codecs.joinToString(",") { it.wireName }.ifBlank {
                    "opus,pcmu,pcma"
                }

            val rc =
                OmnixNative.nativeInit(
                    sipServer = config.sipServer,
                    sipUser = config.sipUser,
                    authUser = config.authUser,
                    sipPassword = config.sipPassword,
                    displayName = config.displayName,
                    stunServer = config.stunServer,
                    verifyCert = config.verifyCert,
                    enableSrtp = config.enableSrtp,
                    codecs = codecsCsv,
                )
            throwIfError(rc)

            initialized = true
            registrationState = OmnixRegistrationState.UNREGISTERED
            Log.i(TAG, "initialized (password not logged)")
        }
    }

    @JvmStatic
    fun shutdown() {
        synchronized(lock) {
            if (!initialized) {
                return
            }
            OmnixNative.nativeSetCallback(null)
            OmnixNative.nativeShutdown()
            calls.clear()
            dispatcher.clear()
            appContext = null
            audioRoute = OmnixAudioRoute.UNKNOWN
            registrationState = OmnixRegistrationState.UNINITIALIZED
            initialized = false
        }
    }

    @JvmStatic
    fun register() {
        synchronized(lock) {
            ensureInitialized()
            throwIfError(OmnixNative.nativeRegister())
        }
    }

    @JvmStatic
    fun unregister() {
        synchronized(lock) {
            ensureInitialized()
            OmnixNative.nativeUnregister()
        }
    }

    @JvmStatic
    fun makeCall(destination: String): String {
        synchronized(lock) {
            ensureInitialized()
            if (destination.isBlank()) {
                throw OmnixVoiceException(
                    OmnixErrorCode.INVALID_CONFIGURATION,
                    "destination is required",
                )
            }
            val callId =
                OmnixNative.nativeMakeCall(destination)
                    ?: throw OmnixVoiceException(
                        OmnixErrorCode.CALL_FAILED,
                        "makeCall failed",
                    )
            return callId
        }
    }

    @JvmStatic
    fun answerCall(callId: String) {
        synchronized(lock) {
            ensureInitialized()
            throwIfError(OmnixNative.nativeAnswerCall(requireCallId(callId)))
        }
    }

    @JvmStatic
    fun rejectCall(callId: String) {
        synchronized(lock) {
            ensureInitialized()
            throwIfError(OmnixNative.nativeRejectCall(requireCallId(callId)))
        }
    }

    @JvmStatic
    fun hangupCall(callId: String) {
        synchronized(lock) {
            ensureInitialized()
            throwIfError(OmnixNative.nativeHangupCall(requireCallId(callId)))
        }
    }

    @JvmStatic
    fun holdCall(callId: String) {
        synchronized(lock) {
            ensureInitialized()
            throwIfError(OmnixNative.nativeHoldCall(requireCallId(callId)))
        }
    }

    @JvmStatic
    fun resumeCall(callId: String) {
        synchronized(lock) {
            ensureInitialized()
            throwIfError(OmnixNative.nativeResumeCall(requireCallId(callId)))
        }
    }

    @JvmStatic
    fun setMuted(callId: String, muted: Boolean) {
        synchronized(lock) {
            ensureInitialized()
            throwIfError(OmnixNative.nativeSetMuted(requireCallId(callId), muted))
        }
    }

    @JvmStatic
    fun setSpeakerEnabled(enabled: Boolean) {
        synchronized(lock) {
            ensureInitialized()
            throwIfError(OmnixNative.nativeSetSpeaker(enabled))
            audioRoute =
                if (enabled) OmnixAudioRoute.SPEAKER else OmnixAudioRoute.EARPIECE
            val route = audioRoute
            dispatcher.dispatch { it.onAudioRouteChanged(route) }
        }
    }

    @JvmStatic
    fun sendDTMF(callId: String, digit: Char) {
        synchronized(lock) {
            ensureInitialized()
            throwIfError(OmnixNative.nativeSendDTMF(requireCallId(callId), digit))
        }
    }

    /**
     * Blind transfer stub until SDK-024. Always throws [OmnixErrorCode.NOT_SUPPORTED].
     */
    @JvmStatic
    fun transferCall(callId: String, destination: String) {
        requireCallId(callId)
        if (destination.isBlank()) {
            throw OmnixVoiceException(
                OmnixErrorCode.INVALID_CONFIGURATION,
                "destination is required",
            )
        }
        throw OmnixVoiceException(
            OmnixErrorCode.NOT_SUPPORTED,
            "blind transfer requires SDK-024",
        )
    }

    @JvmStatic
    fun getCallInfo(callId: String): OmnixCallInfo? = calls[callId]

    @JvmStatic
    fun addListener(listener: OmnixVoiceListener) {
        dispatcher.addListener(listener)
    }

    @JvmStatic
    fun removeListener(listener: OmnixVoiceListener) {
        dispatcher.removeListener(listener)
    }

    private fun ensureInitialized() {
        if (!initialized) {
            throw OmnixVoiceException(
                OmnixErrorCode.INVALID_CALL_STATE,
                "SDK not initialized",
            )
        }
    }

    private fun requireCallId(callId: String): String {
        if (callId.isBlank()) {
            throw OmnixVoiceException(
                OmnixErrorCode.INVALID_CONFIGURATION,
                "callId is required",
            )
        }
        return callId
    }

    private fun throwIfError(code: Int) {
        if (code == NATIVE_OK) {
            return
        }
        val mapped = OmnixErrorCode.fromNative(code)
        throw OmnixVoiceException(mapped, mapped.name)
    }
}

File: ios/Sources/OmnixVoice/OmnixVoice.swift
// Omnix Voice — public Swift facade (Issue #1 §20A.1).
// SDK-041: wraps OmnixVoiceBridge. SDK-042: AVAudioSession via OmnixAudioSession.
// Public API exposes no ObjC / Baresip / re types.
// Placeholders only; never log sipPassword.

import Foundation

public final class OmnixVoice {
    public static let shared = OmnixVoice()

    public weak var delegate: OmnixVoiceDelegate?

    private let lock = NSLock()
    private var audioRoute: OmnixAudioRoute = .UNKNOWN
    private let bridgeSink: BridgeSink
    private let audioSession: OmnixAudioSession
    private var trackedCalls: [String: OmnixCallState] = [:]

    private init() {
        bridgeSink = BridgeSink()
        bridgeSink.owner = self
        audioSession = OmnixAudioSession()
        audioSession.onRouteChanged = { [weak self] route in
            self?.handleExternalRouteChange(route)
        }
        // ObjC `sharedBridge` imports as `shared` in Swift.
        OmnixVoiceBridge.shared.delegate = bridgeSink
    }

    public var registrationState: OmnixRegistrationState {
        lock.lock()
        defer { lock.unlock() }
        let raw = Int(OmnixVoiceBridge.shared.registrationState.rawValue)
        return OmnixRegistrationState.fromBridge(raw)
    }

    public func initialize(config: OmnixVoiceConfig) throws {
        lock.lock()
        defer { lock.unlock() }

        let bridgeConfig = OmnixVoiceBridgeConfig()
        bridgeConfig.sipServer = config.sipServer
        bridgeConfig.sipUser = config.sipUser
        bridgeConfig.sipPassword = config.sipPassword
        bridgeConfig.authUser = config.authUser
        bridgeConfig.displayName = config.displayName
        bridgeConfig.stunServer = config.stunServer
        bridgeConfig.verifyCert = config.verifyCert
        bridgeConfig.enableSrtp = config.enableSrtp
        if !config.codecs.isEmpty {
            bridgeConfig.codecs = config.codecs.map(\.wireName).joined(separator: ",")
        }

        do {
            try OmnixVoiceBridge.shared.initialize(with: bridgeConfig)
        } catch {
            throw mapThrown(error)
        }
    }

    public func shutdown() {
        lock.lock()
        defer { lock.unlock() }
        try? audioSession.endCallAudio()
        OmnixVoiceBridge.shared.shutdown()
        trackedCalls.removeAll()
        audioRoute = .UNKNOWN
    }

    public func register() throws {
        lock.lock()
        defer { lock.unlock() }
        do {
            try OmnixVoiceBridge.shared.register()
        } catch {
            throw mapThrown(error)
        }
    }

    public func unregister() {
        lock.lock()
        defer { lock.unlock() }
        OmnixVoiceBridge.shared.unregister()
    }

    public func makeCall(destination: String) throws -> String {
        lock.lock()
        defer { lock.unlock() }
        var callId: NSString?
        do {
            try OmnixVoiceBridge.shared.makeCall(destination, callId: &callId)
        } catch {
            throw mapThrown(error)
        }
        guard let id = callId as String?, !id.isEmpty else {
            throw OmnixVoiceError(code: .CALL_FAILED, detail: "makeCall failed")
        }
        return id
    }

    public func answerCall(callId: String) throws {
        try invokeCall(callId) { try $0.answerCall($1) }
    }

    public func rejectCall(callId: String) throws {
        try invokeCall(callId) { try $0.rejectCall($1) }
    }

    public func hangupCall(callId: String) throws {
        try invokeCall(callId) { try $0.hangupCall($1) }
    }

    public func holdCall(callId: String) throws {
        try invokeCall(callId) { try $0.holdCall($1) }
    }

    public func resumeCall(callId: String) throws {
        try invokeCall(callId) { try $0.resumeCall($1) }
    }

    public func setMuted(callId: String, muted: Bool) throws {
        try invokeCall(callId) { try $0.setMuted(muted, callId: $1) }
    }

    public func setSpeakerEnabled(_ enabled: Bool) throws {
        lock.lock()
        defer { lock.unlock() }
        do {
            try OmnixVoiceBridge.shared.setSpeakerEnabled(enabled)
            try audioSession.setSpeakerEnabled(enabled)
        } catch let error as OmnixVoiceError {
            throw error
        } catch {
            // AVAudioSession failures surface as MEDIA_ERROR; bridge NSError mapped.
            if (error as NSError).domain == OmnixVoiceBridgeErrorDomain {
                throw mapThrown(error)
            }
            throw OmnixVoiceError(
                code: .MEDIA_ERROR,
                detail: error.localizedDescription
            )
        }
        audioRoute = enabled ? .SPEAKER : .EARPIECE
        let route = audioRoute
        let del = delegate
        DispatchQueue.main.async {
            del?.audioRouteChanged(route)
        }
    }

    public func sendDTMF(callId: String, digit: Character) throws {
        guard let scalar = digit.unicodeScalars.first, scalar.isASCII else {
            throw OmnixVoiceError(
                code: .INVALID_CONFIGURATION,
                detail: "DTMF digit out of range"
            )
        }
        let value = unichar(scalar.value)
        try invokeCall(callId) { try $0.sendDTMF(value, callId: $1) }
    }

    /// Blind transfer stub until SDK-024. Always throws `NOT_SUPPORTED`.
    public func transferCall(callId: String, destination: String) throws {
        if callId.isEmpty || destination.isEmpty {
            throw OmnixVoiceError(
                code: .INVALID_CONFIGURATION,
                detail: "callId and destination are required"
            )
        }
        throw OmnixVoiceError(
            code: .NOT_SUPPORTED,
            detail: "blind transfer requires SDK-024"
        )
    }

    public func callInfo(callId: String) -> OmnixCallInfo? {
        lock.lock()
        defer { lock.unlock() }
        guard let info = OmnixVoiceBridge.shared.callInfo(forCallId: callId) else {
            return nil
        }
        return Self.mapCallInfo(info)
    }

    // MARK: - Internal

    private func invokeCall(
        _ callId: String,
        _ body: (OmnixVoiceBridge, String) throws -> Void
    ) throws {
        lock.lock()
        defer { lock.unlock() }
        if callId.isEmpty {
            throw OmnixVoiceError(
                code: .INVALID_CONFIGURATION,
                detail: "callId is required"
            )
        }
        do {
            try body(OmnixVoiceBridge.shared, callId)
        } catch {
            throw mapThrown(error)
        }
    }

    private func mapThrown(_ error: Error) -> OmnixVoiceError {
        let ns = error as NSError
        let code = OmnixErrorCode.fromNative(ns.code)
        return OmnixVoiceError(code: code, detail: ns.localizedDescription)
    }

    fileprivate static func mapCallInfo(_ info: OmnixVoiceBridgeCallInfo) -> OmnixCallInfo {
        OmnixCallInfo(
            callId: info.callId,
            peerUri: info.peerUri,
            peerDisplayName: info.peerDisplayName,
            state: OmnixCallState.fromBridge(Int(info.state.rawValue)),
            isOutgoing: info.isOutgoing,
            isMuted: info.isMuted,
            isOnHold: info.isOnHold
        )
    }

    fileprivate func emitRegistration(
        state: OmnixRegistrationState,
        sipCode: Int?,
        reason: String?
    ) {
        delegate?.registrationStateChanged(state, sipCode: sipCode, reason: reason)
    }

    fileprivate func emitIncoming(_ call: OmnixCallInfo) {
        rememberCall(call)
        syncCallAudio()
        delegate?.incomingCall(call)
    }

    fileprivate func emitCallState(_ call: OmnixCallInfo) {
        rememberCall(call)
        syncCallAudio()
        delegate?.callStateChanged(call)
    }

    fileprivate func emitError(code: OmnixErrorCode, detail: String?) {
        delegate?.didReceiveError(code, detail: detail, callId: nil)
    }

    private func handleExternalRouteChange(_ route: OmnixAudioRoute) {
        lock.lock()
        audioRoute = route
        lock.unlock()
        delegate?.audioRouteChanged(route)
    }

    /// SDK-042: activate AVAudioSession while any call is non-terminal;
    /// deactivate with notifyOthersOnDeactivation when all calls end.
    private func syncCallAudio() {
        lock.lock()
        let active = trackedCalls.values.contains { Self.isActiveCallState($0) }
        lock.unlock()

        if active {
            do {
                try audioSession.startCallAudio()
            } catch {
                emitError(code: .MEDIA_ERROR, detail: error.localizedDescription)
            }
        } else {
            do {
                try audioSession.endCallAudio()
            } catch {
                emitError(code: .MEDIA_ERROR, detail: error.localizedDescription)
            }
        }
    }

    fileprivate func rememberCall(_ call: OmnixCallInfo) {
        lock.lock()
        trackedCalls[call.callId] = call.state
        lock.unlock()
    }

    private static func isActiveCallState(_ state: OmnixCallState) -> Bool {
        switch state {
        case .OUTGOING, .INCOMING, .RINGING, .EARLY_MEDIA, .CONNECTED, .HELD, .ENDING:
            return true
        case .IDLE, .ENDED, .FAILED:
            return false
        }
    }
}

/// ObjC bridge → Swift delegate adapter (not part of the public API).
private final class BridgeSink: NSObject, OmnixVoiceBridgeDelegate {
    weak var owner: OmnixVoice?

    func omnixVoiceBridge(
        _ bridge: OmnixVoiceBridge,
        registrationStateChanged state: OmnixBridgeRegistrationState,
        sipCode: Int,
        reason: String?
    ) {
        let mapped = OmnixRegistrationState.fromBridge(Int(state.rawValue))
        let code: Int? = sipCode == 0 ? nil : sipCode
        owner?.emitRegistration(state: mapped, sipCode: code, reason: reason)
    }

    func omnixVoiceBridge(_ bridge: OmnixVoiceBridge, incomingCall call: OmnixVoiceBridgeCallInfo) {
        owner?.emitIncoming(OmnixVoice.mapCallInfo(call))
    }

    func omnixVoiceBridge(
        _ bridge: OmnixVoiceBridge,
        callStateChanged call: OmnixVoiceBridgeCallInfo
    ) {
        owner?.emitCallState(OmnixVoice.mapCallInfo(call))
    }

    func omnixVoiceBridge(
        _ bridge: OmnixVoiceBridge,
        error code: OmnixBridgeErrorCode,
        detail: String?
    ) {
        owner?.emitError(code: OmnixErrorCode.fromNative(Int(code.rawValue)), detail: detail)
    }
}
